### ISSUE TITLE
Availability of UniRef validation set

### ISSUE THREAD

**Issue by roman-bushuiev** [evolutionaryscale/esm#107]
Hi! Thank you for the amazing work and the user-friendly codebase!

I would like to reproduce Figure S8, but unfortunately, I couldn’t find the validation set of proteins from UniRef on the [Hugging Face Hub page](https://huggingface.co/EvolutionaryScale/esm3-sm-open-v1/tree/main/data). Is this set please available anywhere, for example as a list of UniProt IDs?

<img width="666" alt="image" src="https://github.com/user-attachments/assets/62cd91fd-7f8f-4e70-b20f-e76890202c13">

<img width="423" alt="image" src="https://github.com/user-attachments/assets/bfcbeb10-2283-4351-8e8f-ce1d36761a49">

**Comment by jondeaton** [comment]
Hi Roman!

Our validation set which has not been released was constructed from UniRef clustered at the 70% sequence identity and purged as per section "A.2.1.7. Purging of Validation Sequences" in the manuscript. The keyword labels used to assess function prediction were made from InterPro annotations from UniProt as per "A.2.1.4. Functional Labels". For this function prediction evaluation, the InterPro annotations were converted to function keyword via the mapping provided in the repository.

For the computation of mean Average Precision (mAP). We computed both a macro average and a micro average using TorchMetrics' [MultilabelAveragePrecision](https://lightning.ai/docs/torchmetrics/stable/classification/average_precision.html#multilabelaverageprecision) and the respective `average` keyword. The predictions and labels are both given on a per-residue basis.

For micro-averaged mAP we computed the mAP for each individual sample in the validation set, then averaged each value uniformly. Computing the macro-averaged mAP involves computing the mAP for each keyword individually over the entire validation set and then averaging the keyword-mAPs uniformly. It is impractical to perform this computation over all ~40k keywords both due to paucity labels for uncommon keywords and memory usage. So we only computed the average over the top 1000 most common keywords, excluding uninformative keywords. I provide the full list here: [top_keywords.txt](https://github.com/user-attachments/files/17403300/top_keywords.txt)

The code that we used to produce the function keywords predictions follows closely the [example](https://github.com/evolutionaryscale/esm/blob/a46ffdc2fc9f91cb6ac30c77d9ddfe89912f6c8a/examples/raw_forwards.py#L121) given under "Function Prediction" in `esm/examples/raw_forwards.py` with one exception:

We noticed that raw function predictions are often "noisy" similar to segmentation boundaries in image segmentation models. We took measures to smooth predictions and this this figure smoothed function-token logits using a Dense Conditional Random Field from [pydensecrf](https://github.com/lucasb-eyer/pydensecrf). I've provided the code to do this below. In the repository we used a simpler smoothing approach of simply dropping short predicted annotations and merging small gaps (see `annotation_min_length` and `annotation_gap_merge_max`), which seemed to have essentially the same effect as the CRF smoothing. We did not observe either smoothing method having a significant impact on final function prediction validation metrics.

```python
import numpy as np
import pydensecrf.densecrf as dcrf
from pydensecrf.utils import create_pairwise_gaussian, unary_from_softmax

def crf_smooth_function_token_logits(
    function_token_logits: torch.Tensor,
    length_scale: float = 3.0,
    compatibility: float = 3.0,
    num_iterations: int = 10,
) -> torch.Tensor:
    """Smooth function token logits with a Conditional Random Field (CRF).

    See https://arxiv.org/abs/1210.5644 for details.

    Args:
        function_token_logits: <float>[length, depth, vocab] function token logits.
        length_scale: affects the width of locality sensitivtiy of the CRF.
        compatibility: parameter affecting smoothness.
        num_iterations: number of CRF iterations to run the model.
    Returns:
        <float>[length, depth, vocab] function token logits processed with the CRF.
    """
    length, depth, vocab = function_token_logits.shape
    p = torch.softmax(function_token_logits, dim=-1).cpu().float().numpy()

    pairwise_energy = create_pairwise_gaussian((length_scale,), (length,))

    crf_logits = torch.zeros_like(function_token_logits)
    for i in range(depth):
        crf = dcrf.DenseCRF(length, vocab)

        unary = unary_from_softmax(p[:, i, :].T).copy()  # (vocab, length)
        crf.setUnaryEnergy(unary)

        crf.addPairwiseEnergy(pairwise_energy, compat=compatibility)
        Q = crf.inference(num_iterations)
        crf_logits[:, i, :] = torch.log(torch.tensor(np.array(Q)).T)

    return crf_logits
```
